### PR TITLE
Add safeguard when resources patch is not ready

### PR DESCRIPTION
**Pull request by michaeldmitry**
## Issue
Intermittent issues in CI where the worker goes into error state during `start` hook.
There could be a race between the resource patch and pebble operations i.e., charm code proceeds beyond a can_connect guard, and then lightkube patches the statefulset and the workload is no longer available.

## Solution
Add a safeguard to not perform pebble operations if the patch is not done.

### CODE REVIEW

File: src/cosl/coordinated_workers/worker.py [hunk 1]
```diff
@@ -406,6 +406,12 @@ def _add_readiness_check(self, new_layer: Layer):
 
     def _reconcile(self):
         """Run all unconditional logic."""
+        # There could be a race between the resource patch and pebble operations
+        # i.e., charm code proceeds beyond a can_connect guard, and then lightkube patches the statefulset
+        # and the workload is no longer available
+        if self.resources_patch and not self.resources_patch.is_ready():
```
Comment (PietroPasotti) [inline src/cosl/coordinated_workers/worker.py:412]:
```suggestion
        if not self.resources_patch.is_ready():
```
or when could the `resources_patch` variable be falsy? please add a comment to explain

File: src/cosl/coordinated_workers/worker.py [hunk 2]
```diff
@@ -406,6 +406,12 @@ def _add_readiness_check(self, new_layer: Layer):
...
+            logger.debug("Resource patching is not yet ready. Skipping config update.")
```
Comment (PietroPasotti) [inline src/cosl/coordinated_workers/worker.py:413]:
```suggestion
            logger.debug("Resource patch not ready yet. Skipping reconciliation step since the pod will shortly be restarted.")
```

Comment (PietroPasotti) [reply src/cosl/coordinated_workers/worker.py:413]: I also think we should surface this in the status. something like "waiting for resource patch to apply". other charms do that too I think

File: src/cosl/coordinated_workers/worker.py [hunk 3]
```diff
@@ -406,6 +406,12 @@ def _add_readiness_check(self, new_layer: Layer):
 
     def _reconcile(self):
         """Run all unconditional logic."""
```
Comment (PietroPasotti) [inline src/cosl/coordinated_workers/worker.py:408]: funny that this says - run all unconditional logic - and the first thing we see two lines below is... a conditional! perhaps rephrase it to say something a bit more precise like "logic that does not depend on what event we're processing"

**Review by PietroPasotti — CHANGES_REQUESTED** [review]
some minor nits, the only thing I really think we should change with this PR is surface this information better to the user by means of a (Waiting) status message.

really awesome that you found the root cause of the issue! I had it in CI once but couldn't repro

File: src/cosl/coordinated_workers/worker.py (hunk 2 above)
Comment (michaeldmitry) [reply src/cosl/coordinated_workers/worker.py:413]: So, that should already happen. `line 277 inside _on_collect_status()`, The status would be WaitingStatus("waiting 
 for bla bla") or BlockedStatus("failed bla bla").

File: src/cosl/coordinated_workers/worker.py (hunk 1 above)
Comment (michaeldmitry) [reply src/cosl/coordinated_workers/worker.py:412]: `resources_patch` can be None when no resources are passed to the worker (don't set requests or limits). Added a comment

File: src/cosl/coordinated_workers/worker.py (hunk 2 above)
Comment (michaeldmitry) [reply src/cosl/coordinated_workers/worker.py:413]: I would change the debug message a bit and not mention "since the pod will shortly be restarted.", as it might have failed. that would be captured in the status message

File: src/cosl/coordinated_workers/worker.py (hunk 3 above)
Comment (michaeldmitry) [reply src/cosl/coordinated_workers/worker.py]: updated docstring

**Review by PietroPasotti — APPROVED** [review]
